Add support for Renesas DA1469x MCU family

andrzej-kaczmarek opened the pull request:
This adds initial support for Renesas DA1469x MCU family.

At the moment following features are supported:
- DA14699 MCU
- DA1469x Development Kit Pro (will work with both DA14695 and DA14699 daughter boards)
- flashing using ezFlashCLI
- GPIO support
- UART support
- flash support (external QSPI connected to QSPIC)

This PR requires extra files that are expected to be stored in separate HAL module (request pending),

andrzej-kaczmarek committed:
soc: Add Renesas SmartBond(tm) DA1469x series SoC

This adds initial support for Renesas SmartBond(tm) DA1469x series SoC.

Initailly only DA14699 is explicitly supported, however this code should
also run on other SoCs in this series since all of them support the same
core features available initially.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>
Signed-off-by: Ben Lauret <ben.lauret.wm@renesas.com>

andrzej-kaczmarek committed:
drivers: gpio: Add driver for smartbond

This adds GPIO driver for Renesas SmartBond(tm) family.

Driver supports pin configuration (input/output) and interrupts on edge.
Interrupts on level are not supported by hardware.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>
Signed-off-by: Ben Lauret <ben.lauret.wm@renesas.com>

andrzej-kaczmarek committed:
drivers: serial: Add driver for smartbond

This adds serial driver for Renesas SmartBond(tm) family. Both polling
and interrupt APIs are supported.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>
Signed-off-by: Ben Lauret <ben.lauret.wm@renesas.com>

andrzej-kaczmarek committed:
drivers: flash: Add driver for smartbond

This adds flash driver for Renesas SmartBond(tm) family.

This technically uses QSPI controller but since default and most
commonly used configuration is to boot from external QSPI flash (DA1469x
do not have built-in flash) and that flash is mapped into memory space,
it can be represented as internal flash.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>
Signed-off-by: Ben Lauret <ben.lauret.wm@renesas.com>

andrzej-kaczmarek committed:
boards: Add DA1469x Pro DK

This adds support for DA1469x Pro DK which is an official development
kit for DA1469x SoC series provided by Renesas.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>
Signed-off-by: Ben Lauret <ben.lauret.wm@renesas.com>

andrzej-kaczmarek committed:
scripts: Add ezFlashCLI runner

This adds support for ezFlashCLI tool to allow flashing DA1469x MCU
family from Renesas.

The tool is available at https://github.com/ezflash/ezFlashCLI or can
be installed via pip.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>
Signed-off-by: Ben Lauret <ben.lauret.wm@renesas.com>

andrzej-kaczmarek commented:
@blauret	@NiekIlmer FYI

carlescufi commented:
@andrzej-kaczmarek thanks for the PR. Can you please also point `west.yml` to the new `hal_renesas`? Otherwise I assume this will not build? Just point to a repo in your own user org for now, then we will move it to `zephyrproject-org`.

andrzej-kaczmarek committed:
west.yml: Add hal_renesas module

This adds hal_renesas module to allow builds for Renesas DA1469x.
For now this points to repository outside zephyrproject-rtos since
relevant external module is not yet created.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

gmarull commented on boards/arm/da1469x_dk_pro/da1469x_dk_pro.dts:
```diff
@@ -72,3 +74,10 @@
+&uart {
+	status = "okay";
+	current-speed = <115200>;
+	tx-pin = <9>;
+	rx-pin = <8>;
```
this is no longer allowed, please, create a pinctrl driver, ref. https://docs.zephyrproject.org/latest/hardware/pinctrl/index.html

andrzej-kaczmarek commented:
> @andrzej-kaczmarek thanks for the PR. Can you please also point `west.yml` to the new `hal_renesas`? Otherwise I assume this will not build? Just point to a repo in your own user org for now, then we will move it to `zephyrproject-org`.

@carlescufi done, all green now

carlescufi commented:
> > @andrzej-kaczmarek thanks for the PR. Can you please also point `west.yml` to the new `hal_renesas`? Otherwise I assume this will not build? Just point to a repo in your own user org for now, then we will move it to `zephyrproject-org`.
> 
> @carlescufi done, all green now

Thanks, looks good!

carlescufi commented on boards/arm/da1469x_dk_pro/da1469x_dk_pro.yaml:
```diff
@@ -0,0 +1,11 @@
+arch: arm
+toolchain:
+  - gnuarmemb
+  - zephyr
```
can you move this to the top? the Zephyr SDK is now the "default" toolchain

carlescufi reviewed (changes requested):
Blocking until @nvlsianpu has had a chance to look at the flash driver.

nandojve reviewed (changes requested):
Hi @andrzej-kaczmarek ,

Thank you for your PR. I would like continue review after you reorder commits by dependencies.

andrzej-kaczmarek commented:
I think I fixed all the comments except for the UART pins - I need to finish pinctrl driver to fix that (hopefully soon), but in case this is the only issue left after next rounds of review I'd like to just remove it from this PR and deliver in separate on with pinctrl driver, if that's ok.

The installation instructions for ezFlashCli are in board documentation since those are only relevant for this board.

ydamigos commented on soc/arm/renesas_da1469x/Kconfig.soc:
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2022 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_FAMILY_DA1469X
```
The family/line is Renesas' SmartBond and SoC series is DA1469x, isn't it?

aaillet replied:
@andrzej-kaczmarek If you confirm this family name, please rename drivers + bindings, please also rework file tree for `dts/arm/renesas/smartbond` as I'm gonna add a `dts/arm/renesas/rcar` folder to store renesas rcar device trees !

andrzej-kaczmarek replied:
@aaillet I'll change family name to `smartbond` and series to `da1469x` and also update driver names to `..._smartbond` as well

ydamigos commented on soc/arm/renesas_da1469x/da14699/linker.ld:
```diff
@@ -0,0 +1,7 @@
+/*
+ * Copyright (c) 2014 Wind River Systems, Inc.
```
```suggestion
 * Copyright (c) 2022 Renesas Electronics Corporation
```

aaillet commented on boards/arm/da1469x_dk_pro/da1469x_dk_pro.dts:
```diff
@@ -0,0 +1,115 @@
+	status = "okay";
+};
+
+&uart {
```
Please document which physical port is providing this bus in board documentation

aaillet commented:
Did you validated overall operation by running the whole twister test suite on the board locally ?

gmarull commented:
> I think I fixed all the comments except for the UART pins - I need to finish pinctrl driver to fix that (hopefully soon), but in case this is the only issue left after next rounds of review I'd like to just remove it from this PR and deliver in separate on with pinctrl driver, if that's ok.
> 
> The installation instructions for ezFlashCli are in board documentation since those are only relevant for this board.

Pin control driver should be part of this PR, there's no reason to merge something that is going to be replaced soon (and that is no longer allowed).

ydamigos commented on scripts/west_commands/runners/ezflashcli.py:
```diff
@@ -0,0 +1,54 @@
+    def do_run(self, command, **kwargs):
+        self.require(self.tool)
+        self.ensure_output('bin')
+        self.program_bin()
```
You could reset target after flashing the image so the application can run immediately without waiting for external reset.
It will be helpful for running twister.

galak commented on drivers/flash/Kconfig.da1469x:
```diff
@@ -0,0 +1,12 @@
+	select FLASH_HAS_PAGE_LAYOUT
+	select FLASH_HAS_DRIVER_ENABLED
+	select MPU_ALLOW_FLASH_WRITE if ARM_MPU
+	default y
```
make this based on `default $(dt_compat_enabled,$(DT_COMPAT_RENESAS_DA1469X_FLASH_CONTROLLER))`

galak commented on drivers/gpio/Kconfig.da1469x:
```diff
@@ -0,0 +1,9 @@
+
+config GPIO_DA1469X
+	bool "DA1469x GPIO driver"
+	default y
```
make this based on default `$(dt_compat_enabled,$(DT_COMPAT_RENESAS_DA1469X_GPIO))`

galak commented on dts/arm/renesas/da1469x/da1469x.dtsi:
```diff
@@ -0,0 +1,79 @@
+			       0x50000070 12
+			       0x50000114 36>;
+			interrupts = <38 0>;
+			label = "GPIO_0";
```
remove `label` property where not needed from dts

andrzej-kaczmarek committed:
drivers: pinctrl: Add driver for smartbond

This adds pinctrl driver for Renesas SmartBond(tm) MCU family.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek commented:
All issues should be now fixed.

MCU family is changed to `smartbond` and SoC series is `da1469x`.
pinctrl driver is also added, UART is configured accordingly.
Target will be now reset after flashing.

andrzej-kaczmarek commented:
> > All issues should be now fixed.
> > MCU family is changed to `smartbond` and SoC series is `da1469x`. pinctrl driver is also added, UART is configured accordingly. Target will be now reset after flashing.
> 
> Please also rename `da1469x` folder in `hal_renesas` for `smartbond` !

ah, forgot to commit those changes. fixed now!

nandojve commented on soc/arm/renesas_smartbond/Kconfig.defconfig:
```diff
@@ -0,0 +1,11 @@
+
+source "soc/arm/renesas_smartbond/*/Kconfig.defconfig.series"
+
+#config SYS_CLOCK_HW_CYCLES_PER_SEC
+#	default 32000000
```
This should be picked from devicetree as below: https://github.com/zephyrproject-rtos/zephyr/blob/c699ae5850297c824f8a7ca2a4f06174a9782fd3/soc/arm/gigadevice/gd32f4xx/Kconfig.defconfig.gd32f470#L9-L10

BTW, is this the CPU core frequency or just crystal value?

nandojve commented on soc/arm/renesas_smartbond/da1469x/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,17 @@
+config NUM_IRQS
+	default 40
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 32000000
```
Why there is two places for this ?

see `soc/arm/renesas_smartbond/Kconfig.defconfig`

nandojve commented on boards/arm/da1469x_dk_pro/doc/index.rst:
```diff
@@ -0,0 +1,119 @@
+| RTT       | Segger     | console              |
++-----------+------------+----------------------+
+| UART      | on-chip    | serial               |
++-----------+------------+----------------------+
+| WDT       | on-chip    | watchdog             |
```
Where is WDT driver?
Why not mention pinctrl driver here?

andrzej-kaczmarek replied:
Watchdog is only initialized by SoC code, there's no dedicated driver (yet) - I'll just remove entry for now.
I don't really see pinctrl driver mentioned in any of board docs...

nandojve commented on dts/arm/renesas/smartbond/da14699.dtsi:
```diff
@@ -0,0 +1,7 @@
+#include "da1469x.dtsi"
+
+&sram0 {
+	reg = <0x20000200 (DT_SIZE_K(512) - 0x200)>;
```
Why? Maybe you should add a note to make people aware of it.

andrzej-kaczmarek replied:
first 512 bytes of sram have to be reserved because of how memory remapping is configured after reset, but I managed to change this to be handled in linker script so dt regs are not confusing anymore

nandojve commented on boards/arm/da1469x_dk_pro/da1469x_dk_pro.yaml:
```diff
@@ -0,0 +1,12 @@
+name: DA1469X_DK_PRO
+type: mcu
+arch: arm
+ram: 512
```
Based on your SoC devicetree this value should be different.

nandojve reviewed (dismissed):
Hi @andrzej-kaczmarek ,

I made another pass on configurations, a few more comments.

aaillet commented on boards/arm/da1469x_dk_pro/Kconfig:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2022 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+if BOARD_DA1469X_DK_PRO
```
Is this file really needed if not used to enable anything ?

ydamigos commented on drivers/gpio/gpio_smartbond.c:
```diff
@@ -0,0 +1,278 @@
+	static const struct gpio_smartbond_config gpio_smartbond_p##id##_config = {	\
+		.common = {								\
+			.port_pin_mask =						\
+			GPIO_PORT_PIN_MASK_FROM_NGPIOS(id),				\
```
I don't think this is correct. I would expect GPIO_PORT_PIN_MASK_FROM_DT_INST(id).

ydamigos reviewed (changes requested):
Running twister locally on board 16 test cases fail.
`./scripts/twister --device-testing --device-serial /dev/ttyUSB0 --device-serial-baud 115200 -p da1469x_dk_pro`

Did you check them? One is related to the comment I left in the GPIO driver.

gmarull commented on soc/arm/renesas_smartbond/da1469x/soc.c:
```diff
@@ -0,0 +1,42 @@
+#include <init.h>
+#include <arch/arm/aarch32/cortex_m/cmsis.h>
+
+#ifdef CONFIG_RUNTIME_NMI
+extern void z_arm_nmi_init(void);
+#define NMI_INIT() z_arm_nmi_init()
+#else
+#define NMI_INIT()
+#endif
```
seems to be declared in include/zephyr/arch/arm/aarch32/nmi.h ?

gmarull commented on soc/arm/renesas_smartbond/da1469x/soc.h:
```diff
@@ -0,0 +1,26 @@
+extern "C" {
+#endif
+
+#include <stdint.h>
+
+#include <devicetree.h>
+#include <sys/util.h>
+#include "DA1469xAB.h"
```
these are likely not needed, please remove. `soc.h` needs to exist on ARM because certain MPU code relies on it for indirect dependencies (needs to be fixed). Just include the bare minimum for that case.

gmarull commented on soc/arm/renesas_smartbond/da1469x/soc.h:
```diff
@@ -0,0 +1,26 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _SOC__H_
+#define _SOC__H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+
+#include <devicetree.h>
+#include <sys/util.h>
+#include "DA1469xAB.h"
+
+#ifdef __cplusplus
+}
+#endif
+
+#include <core_cm33.h> /*!< ARM Cortex-M33 processor and core peripherals */
```
this file just seem to need `#include <DA1469xAB.h>`, all others are not required (and we need to fix such indirect dependency at some point, so we can finally remove soc.h on ARM).

gmarull commented on drivers/gpio/gpio_smartbond.c:
```diff
@@ -0,0 +1,278 @@
+#define DT_DRV_COMPAT renesas_smartbond_gpio
+
+#include <stdint.h>
+#include <drivers/gpio.h>
```
please use `<zephyr/>` prefix everywhere

gmarull commented on drivers/gpio/gpio_smartbond.c:
```diff
@@ -0,0 +1,278 @@
+	}
+
+	if (flags & GPIO_OUTPUT) {
+		config->mode_regs[pin] = 0x300;
+
+		if (flags & GPIO_OUTPUT_INIT_LOW) {
+			config->data_regs->reset = 1 << pin;
+		} else if (flags & GPIO_OUTPUT_INIT_HIGH) {
+			config->data_regs->set = 1 << pin;
+		}
+
+		return 0;
+	}
+
+	if (flags & GPIO_PULL_DOWN) {
+		config->mode_regs[pin] = 0x200;
+	} else if (flags & GPIO_PULL_UP) {
+		config->mode_regs[pin] = 0x100;
+	} else {
+		config->mode_regs[pin] = 0x000;
```
it would be nice to add defines for all such magic constants

gmarull commented on drivers/gpio/gpio_smartbond.c:
```diff
@@ -0,0 +1,278 @@
+		config->mode_regs[pin] = 0x300;
+
+		if (flags & GPIO_OUTPUT_INIT_LOW) {
+			config->data_regs->reset = 1 << pin;
+		} else if (flags & GPIO_OUTPUT_INIT_HIGH) {
+			config->data_regs->set = 1 << pin;
```
`BIT(pin)`

gmarull commented on drivers/gpio/gpio_smartbond.c:
```diff
@@ -0,0 +1,278 @@
+{
+	const struct gpio_smartbond_config *config = dev->config;
+
+	if (flags == GPIO_DISCONNECTED) {
+		return 0;
+	}
```
this option should actually disable/disconnect the gpio (if applicable, e.g. to save power)

gmarull commented on drivers/gpio/gpio_smartbond.c:
```diff
@@ -0,0 +1,278 @@
+	.manage_callback = gpio_smartbond_manage_callback,
+};
+
+#define GPIO_SMARTBOND_CONFIGURE(id)			\
+	do {						\
+		IRQ_CONNECT(DT_INST_IRQN(id),		\
+			    DT_INST_IRQ(id, priority),	\
+			    gpio_smartbond_isr,		\
+			    DEVICE_DT_INST_GET(id), 0);	\
+							\
+		irq_enable(DT_INST_IRQN(id));		\
+	} while (0)
```
why not place this under `gpio_smartbond_##id##_init`?

gmarull commented on drivers/pinctrl/pinctrl_smartbond.c:
```diff
@@ -0,0 +1,56 @@
+#include <zephyr/drivers/pinctrl.h>
+#include <DA1469xAB.h>
+
+/** Utility macro to retrieve starting mode register and pin count for GPIO port from DT */
+#define GPIO_PORT_ENTRY(nodelabel)							\
+		{ DT_REG_ADDR_BY_IDX(DT_NODELABEL(nodelabel), 1),			\
+		  DT_REG_SIZE_BY_IDX(DT_NODELABEL(nodelabel), 1) / sizeof(uint32_t) }
+
+struct gpio_port {
+    uint32_t p0_mode_addr;
+    uint8_t pin_count;
+};
+
+static const struct gpio_port smartbond_gpio_ports[] = {
+	GPIO_PORT_ENTRY(gpio0),
+	GPIO_PORT_ENTRY(gpio1),
+};
```
please add gpio driver first, this is not bisectable as these nodes don't exist yet

gmarull commented on dts/arm/renesas/smartbond/da1469x.dtsi:
```diff
@@ -21,12 +22,36 @@
+			port = <0>;
+		};
+
+		gpio1: gpio@50020a04 {
```
seems like pinctrl is re-using gpio regs, so you should probably do like STM32/GD32:

```
pinctrl: pin-controller@... {
     ...
     gpio0: gpio@... {
     };
};
```

gmarull commented on drivers/serial/Kconfig.smartbond:
```diff
@@ -0,0 +1,12 @@
+
+config UART_SMARTBOND
+	bool "Renesas SmartBond(tm) UART driver"
+	default y
```
please default based on dt compat

gmarull commented on drivers/serial/uart_smartbond.c:
```diff
@@ -0,0 +1,506 @@
+			      UART2_UART2_SRR_REG_UART_XFR_Msk;
+
+	config->regs->UART2_LCR_REG |= UART2_UART2_LCR_REG_UART_DLAB_Msk;
+	config->regs->UART2_IER_DLH_REG = (baudrate_cfg >> 16) & 0xff;
+	config->regs->UART2_RBR_THR_DLL_REG = (baudrate_cfg >> 8) & 0xff;
```
`zephyr/byteops.h` API?

gmarull replied:
sorry, it's `zephyr/sys/byteorder.h`

gmarull commented on drivers/serial/uart_smartbond.c:
```diff
@@ -0,0 +1,506 @@
+
+	while (!no_intr) {
+		switch (config->regs->UART2_IIR_FCR_REG & 0x0F) {
+		case 0x01:
+			/* no interrupt pending */
+			no_intr = true;
+			break;
+		case 0x02:
+			/* THR empty */
+			irq_tx_disable(dev);
+			break;
+		case 0x04:
+			/* received data available */
+			irq_rx_disable(dev);
+			break;
+		case 0x06:
+			/* receiver line status */
+		case 0x0c:
+			/* character timeout */
```
define magic constants?

gmarull commented on drivers/serial/uart_smartbond.c:
```diff
@@ -0,0 +1,506 @@
+		case 0x0c:
+			/* character timeout */
+			break;
+		case 0x07:
+			/* busy detect */
```
fall-through intentional here? (leads to a panic)

gmarull commented on drivers/serial/uart_smartbond.c:
```diff
@@ -0,0 +1,506 @@
+		return -ENOTSUP;
+	}
+
+	/* Flow control is not supported on UART */
```
not supported on `uart0` I guess? if that's the case, it would be better to specify this in DT, e.g. using a boolean option `flow-control-supported;` in nodes that support it. See for example https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/gigadevice/gd32f4xx/gd32f4xx.dtsi#L367 or https://github.com/zephyrproject-rtos/zephyr/pull/46984/commits/2e7304abea25d134ca5ba937f841f685c52b7fb8

gmarull reviewed (changes requested):
thanks for the continued effort, please check my comments.

andrzej-kaczmarek commented on drivers/flash/flash_smartbond.c:
```diff
@@ -0,0 +1,286 @@
+#define FLASH_PAGE_SIZE		256
+
+static const struct flash_parameters flash_smartbond_parameters = {
+	.write_block_size = DT_PROP(SOC_NV_FLASH_NODE, write_block_size),
```
this is generic structure used by pretty much every SoC flash driver

gmarull replied:
right, I got confused, sorry.

andrzej-kaczmarek commented on drivers/serial/uart_smartbond.c:
```diff
@@ -0,0 +1,506 @@
+
+	key = k_spin_lock(&data->lock);
+
+	switch (config->periph_id) {
+	case 0:
+		CRG_COM->SET_CLK_COM_REG = CRG_COM_SET_CLK_COM_REG_UART_ENABLE_Msk;
+		break;
+	case 1:
+		CRG_COM->SET_CLK_COM_REG = CRG_COM_SET_CLK_COM_REG_UART2_ENABLE_Msk;
+		break;
+	case 2:
+		CRG_COM->SET_CLK_COM_REG = CRG_COM_SET_CLK_COM_REG_UART3_ENABLE_Msk;
+		break;
+	default:
+		return -EINVAL;
+	}
```
I will add clock control interface at some point, but would prefer to have basic functionality that is already working merged first

andrzej-kaczmarek commented on dts/bindings/serial/renesas,smartbond-uart.yaml:
```diff
@@ -0,0 +1,38 @@
+    interrupts:
+      required: true
+
+    peripheral-id:
+      type: int
+      description: |
+        UART peripheral identifier. 0 for UART, 1 for UART2 and 2 for UART3.
+      required: true
+
```
I don't think this is necessary since there are 3 physical instances of UART block, but there's only single GPIO block with pins grouped into 2 ports - p0 and p1 - so there are 2 nodes in device tree for each port

andrzej-kaczmarek commented:
> Running twister locally on board 16 test cases fail. `./scripts/twister --device-testing --device-serial /dev/ttyUSB0 --device-serial-baud 115200 -p da1469x_dk_pro`
> 
> Did you check them? One is related to the comment I left in the GPIO driver.

I fixed GPIO failures, there are some "no console timeout" failures but I am not quite sure what is wrong since the same apps work just fine outside twister...

ydamigos commented on drivers/gpio/gpio_smartbond.c:
```diff
@@ -0,0 +1,286 @@
+	static const struct gpio_smartbond_config gpio_smartbond_p##id##_config = {	\
+		.common = {								\
+			.port_pin_mask =						\
+			GPIO_PORT_PIN_MASK_FROM_DT_INST(id),				\
```
`GPIO_PORT_PIN_MASK_FROM_DT_INST()` is defined as:
```
#define GPIO_PORT_PIN_MASK_FROM_DT_INST(inst)                   \                                            
             GPIO_PORT_PIN_MASK_FROM_NGPIOS(DT_INST_PROP(inst, ngpios))
```
I couldn't find ngpios property in SoC's dts file.

ydamigos commented on drivers/pinctrl/pinctrl_smartbond.c:
```diff
@@ -0,0 +1,56 @@
+/** Utility macro to retrieve starting mode register and pin count for GPIO port from DT */
+#define GPIO_PORT_ENTRY(nodelabel)							\
+		{ DT_REG_ADDR_BY_IDX(DT_NODELABEL(nodelabel), 1),			\
+		  DT_REG_SIZE_BY_IDX(DT_NODELABEL(nodelabel), 1) / sizeof(uint32_t) }
```
I am not sure about it. For DA14695, the registers are there but the pins aren't due to package.
gpio node's `ngpios` property looks like a better option.

mbolivar-nordic commented on dts/bindings/flash_controller/renesas,smartbond-flash-controller.yaml:
```diff
@@ -0,0 +1,14 @@
+include: flash-controller.yaml
+
+properties:
+  qspif_base_address:
```
Use dashes (`-`), not underscores (`_`), to separate words in properties.

mbolivar-nordic commented on dts/bindings/flash_controller/renesas,smartbond-flash-controller.yaml:
```diff
@@ -0,0 +1,14 @@
+  qspif_base_address:
+    type: int
+    required: true
+    description: Base address of QSPIF in memory map
```
This looks strange. Why not just take a phandle to the QSPIF so you can use DT_REG_ADDR on that node?

andrzej-kaczmarek replied:
The "flash" node in device tree is located at 0x0 because this is where internal bootloader remaps it after boot so interrupt vector is at 0x0. Flash controller has to access memory at 0x16000000 instead of 0x0 so this is the offset (i.e. qspif-base-address) I need here. If I set "flash" node at 0x16000000, the VTOR will be also set to 0x16000000 which is ro, so any attempt to modify interrupt vector will result in a fault.

So not sure what you mean by taking a handle to QSPIF - should I add a separate QSPIF node to device tree?

gmarull replied:
QSPIF can be described in Devicetree, so you can take its address from its node.

mbolivar-nordic commented on dts/bindings/serial/renesas,smartbond-uart.yaml:
```diff
@@ -0,0 +1,43 @@
+    interrupts:
+      required: true
+
+    peripheral-id:
+      type: int
+      description: |
+        UART peripheral identifier. 0 for UART, 1 for UART2 and 2 for UART3.
```
Same issue here -- HAL limitation workaround, or actually programmed into a register?

andrzej-kaczmarek replied:
This is used to enable clock for proper instance. I found at least few other drivers doing exactly the same thing so not sure why it's not ok here?

mbolivar-nordic replied:
Sometimes things get merged that misbehave. There are too many devicetree bindings for the DT maintainers to review, and not all subsystem maintainers catch things like this.

mbolivar-nordic commented on scripts/west_commands/runners/ezflashcli.py:
```diff
@@ -0,0 +1,59 @@
+            self.logger.info("Erasing flash...")
+            self.check_call([self.tool, "erase_flash"])
+
+        self.logger.info("Flashing...")
```
Suggestion:

```
         self.logger.info(f"Flashing {self.bin_}...")
```

mbolivar-nordic commented on scripts/west_commands/runners/ezflashcli.py:
```diff
@@ -0,0 +1,59 @@
+        self.bin_ = cfg.bin_file
+
+        self.tool = tool
+        self.sn = sn
```
This is unused, which looks like a bug.

andrzej-kaczmarek commented:
added `ngpios` to GPIO nodes
removed `port` from GPIO binding (not needed)
fixed `sn` in ezCliFlash runner
replaced `peripheral-id` in serial driver with clock configuration property - this is actually written into clock configuration register

I am not sure it last change is ok now and whether offset in flash controller should be changed - if yes, any tip on how to implement this properly would be really helpful :)

gmarull commented on soc/arm/renesas_smartbond/da1469x/soc.c:
```diff
@@ -0,0 +1,35 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <init.h>
```
`<zephyr/`> (CI should fail now if you rebase)

gmarull commented on drivers/gpio/Kconfig.smartbond:
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2022 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_RENESAS_SMARTBOND_GPIO := renesas,smartbond-gpio
+
+config GPIO_SMARTBOND
+	bool "Renesas SmartBond(tm) GPIO driver"
+	default $(dt_compat_enabled,$(DT_COMPAT_RENESAS_SMARTBOND_GPIO))
+	depends on SOC_FAMILY_SMARTBOND
```
please update to use `DT_HAS*` helpers (see other GPIO drivers)

gmarull commented on drivers/gpio/gpio_smartbond.c:
```diff
@@ -0,0 +1,284 @@
+
+#define DT_DRV_COMPAT renesas_smartbond_gpio
+
+#include <stdint.h>
+#include <zephyr/drivers/gpio.h>
+#include <DA1469xAB.h>
+#include "gpio_utils.h"
```
```
#include "gpio_utils.h"

#include <stdint.h>
#include <zephyr/drivers/gpio.h>

#include <DA1469xAB.h>
```

andrzej-kaczmarek replied:
I cannot change order as proposed as this will not build - `gpio_utils.h` does not include required headers, e.g. `stdint.h`

gmarull replied:
FYI https://github.com/zephyrproject-rtos/zephyr/pull/49118

gmarull commented on drivers/gpio/gpio_smartbond.c:
```diff
@@ -0,0 +1,284 @@
+			.port_pin_mask =						\
+			GPIO_PORT_PIN_MASK_FROM_DT_INST(id),				\
+		},									\
+		.data_regs = (volatile void *)DT_INST_REG_ADDR_BY_IDX(id, 0),		\
+		.mode_regs = (volatile void *)DT_INST_REG_ADDR_BY_IDX(id, 1),		\
+		.latch_regs = (volatile void *)DT_INST_REG_ADDR_BY_IDX(id, 2),		\
+		.wkup_regs = (volatile void *)DT_INST_REG_ADDR_BY_IDX(id, 3),		\
```
suggestion here: use `reg-names` in DT and retrieve by name instead.

gmarull commented on dts/bindings/gpio/renesas,smartbond-gpio.yaml:
```diff
@@ -0,0 +1,19 @@
+
+gpio-cells:
+  - pin
+  - flags
```
note: please confirm `tests/drivers/gpio` pass

gmarull commented on drivers/pinctrl/Kconfig.smartbond:
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2022 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_SMARTBOND_PINCTRL := renesas,smartbond-pinctrl
+
+config PINCTRL_SMARTBOND
+	bool "Renesas SmartBond(tm) pinctrl driver"
+	depends on SOC_FAMILY_SMARTBOND
+	default $(dt_compat_enabled,$(DT_COMPAT_SMARTBOND_PINCTRL))
```
ditto, use `DT_HAS*`

gmarull commented on drivers/serial/Kconfig.smartbond:
```diff
@@ -0,0 +1,14 @@
+config UART_SMARTBOND
+	bool "Renesas SmartBond(tm) UART driver"
+	default $(dt_compat_enabled,$(DT_COMPAT_RENESAS_SMARTBOND_UART))
+	select PINCTRL
```
regarding pinctrl: better enable it at SoC level (soc `Kconfig.defconfig` file, see e.g. GD32).

gmarull commented on drivers/serial/Kconfig.smartbond:
```diff
@@ -0,0 +1,14 @@
+	select PINCTRL
+	select SERIAL_HAS_DRIVER
+	select SERIAL_SUPPORT_INTERRUPT
+	depends on SOC_FAMILY_SMARTBOND
```
this dependency is redundant if using `DT_HAS*` helpers

gmarull commented on dts/arm/renesas/smartbond/da1469x.dtsi:
```diff
@@ -55,6 +55,15 @@
+		uart: uart@50020000 {
+			compatible = "renesas,smartbond-uart";
+			reg = <0x50020000 0x100>;
+			periph-clock-config = <0x01>;
```
for future: you should implement a proper clock control driver and use standard `clocks` property.

gmarull commented on dts/arm/renesas/smartbond/da1469x.dtsi:
```diff
@@ -55,6 +55,15 @@
+			periph-clock-config = <0x01>;
+			interrupts = <5 0>;
+			status = "disabled";
+			label = "UART";
```
please remove all labels, they're no longer required.

gmarull commented on drivers/serial/uart_smartbond.c:
```diff
@@ -0,0 +1,501 @@
+#include <zephyr/spinlock.h>
+#include <DA1469xAB.h>
+
+#define DT_DRV_COMPAT renesas_smartbond_uart
```
by convention: this goes to the top

gmarull commented on drivers/serial/uart_smartbond.c:
```diff
@@ -0,0 +1,501 @@
+
+	switch (cfg->data_bits) {
+	case UART_CFG_DATA_BITS_6:
+		reg_val |= 1 << UART2_UART2_LCR_REG_UART_DLS_Pos;
+		break;
+	case UART_CFG_DATA_BITS_7:
+		reg_val |= 2 << UART2_UART2_LCR_REG_UART_DLS_Pos;
+		break;
+	case UART_CFG_DATA_BITS_8:
+		reg_val |= 3 << UART2_UART2_LCR_REG_UART_DLS_Pos;
```
nit: add definitions for `1 << UART2_UART2_LCR_REG_UART_DLS_Pos` and others?

gmarull commented on drivers/serial/uart_smartbond.c:
```diff
@@ -0,0 +1,501 @@
+
+	/* Enable hardware FIFO */
+	config->regs->UART2_SFE_REG = UART2_UART2_SFE_REG_UART_SHADOW_FIFO_ENABLE_Msk;
+	config->regs->UART2_SRT_REG = 0;
+	config->regs->UART2_STET_REG = 3;
```
magic numbers?

gmarull commented on drivers/serial/uart_smartbond.c:
```diff
@@ -0,0 +1,501 @@
+		return -ENOTSUP;
+	}
+
+	key = k_spin_lock(&data->lock);
```
question: why is spinlock required here?

andrzej-kaczmarek replied:
registers that are used for configuration have different function during normal usage, thus they are protected here and in other places that access them (e.g. FIFO access)

gmarull commented on drivers/flash/Kconfig.smartbond:
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2022 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_RENESAS_SMARTBOND_FLASH_CONTROLLER := renesas,smartbond-flash-controller
+
+config SOC_FLASH_SMARTBOND
+	bool "Renesas SmartBond(tm) flash driver"
+	depends on SOC_FAMILY_SMARTBOND
+	select FLASH_HAS_PAGE_LAYOUT
+	select FLASH_HAS_DRIVER_ENABLED
+	select MPU_ALLOW_FLASH_WRITE if ARM_MPU
+	default $(dt_compat_enabled,$(DT_COMPAT_RENESAS_SMARTBOND_FLASH_CONTROLLER))
```
ditto `DT_HAS*`

gmarull commented on boards/arm/da1469x_dk_pro/da1469x_dk_pro_defconfig:
```diff
@@ -0,0 +1,18 @@
+CONFIG_UART_CONSOLE=y
+CONFIG_SERIAL=y
+
+CONFIG_USE_SEGGER_RTT=y
```
is this always required?

gmarull reviewed (commented):
a few more comments, but IMO this is in pretty good shape now.

carlescufi commented:
> @andrzej-kaczmarek thanks for the PR. Can you please also point `west.yml` to the new `hal_renesas`? Otherwise I assume this will not build? Just point to a repo in your own user org for now, then we will move it to `zephyrproject-org`.

@andrzej-kaczmarek My apologies for the delay with your work. Can you please now point to https://github.com/zephyrproject-rtos/hal_renesas

carlescufi commented:
@gmarull sorry did not want to dismiss your review.

gmarull commented:
> @gmarull sorry did not want to dismiss your review.

it's not dismissed, it still counts as changes requested.

andrzej-kaczmarek commented:
all comments should be fixed now.

I added simple node for qspif with regs only to use it in flash controller to get proper offset for accessing flash.
test/drviers/gpio pass.
Regarding byteorder.h in uart_smartbond.c (for some reason I cannot reply to that particular comment directly...) - those macros are not really useful since I don't need to handle byte order swapping, just need to extract bytes from configuration value; I added helper macros instead to improve readability
also west.yml points now to hal_renesas repo

gmarull commented on drivers/gpio/Kconfig.smartbond:
```diff
@@ -0,0 +1,8 @@
+
+config GPIO_SMARTBOND
+	bool "Renesas SmartBond(tm) GPIO driver"
+	default DT_HAS_RENESAS_SMARTBOND_GPIO_ENABLED
```
```
default y
depends on DT_HAS_RENESAS_SMARTBOND_GPIO_ENABLED
```

gmarull commented on drivers/flash/Kconfig.smartbond:
```diff
@@ -0,0 +1,12 @@
+	select FLASH_HAS_PAGE_LAYOUT
+	select FLASH_HAS_DRIVER_ENABLED
+	select MPU_ALLOW_FLASH_WRITE if ARM_MPU
+	default DT_HAS_RENESAS_SMARTBOND_FLASH_CONTROLLER_ENABLED
```
ditto

andrzej-kaczmarek commented:
fixed Kconfig issues

gmarull commented on dts/arm/renesas/smartbond/da1469x.dtsi:
```diff
@@ -24,6 +24,24 @@
 		};
 
+		qspif: memory@16000000 {
+		    reg = <0x16000000 DT_SIZE_K(32768)>;
```
last nit: mixes tabs/spaces, this is also missing a compatible string.

ydamigos reviewed (approved):
Looks good to me.

carlescufi commented:
@mbolivar-nordic @nandojve I've dismissed your reviews. Please take another look.

aaillet reviewed (approved):
Everything LGTM now !
I'll work on renesas file tree soon to put both rcar and smartbond familys at the same level soon !
Thanks for your work

nandojve reviewed (approved):
LGTM
